falcon/cluster/cluster.py

```diff
@@ -145,93 +180,122 @@ def _build_query_ann_index(
-        for batch_start in range(0, n_split, batch_size):
-            batch_stop = min(batch_start + batch_size, n_split)
-            # noinspection PyArgumentList
-            index.add_with_ids(vectors_split[batch_start:batch_stop],
-                               np.arange(batch_start, batch_stop,
-                                         dtype=np.int64))
-        # Query the index to calculate NN distances.
-        _dist_mz_interval(
-            index, vectors_split, precursor_mzs[-1], rts[-1], batch_size,
-            n_neighbors, n_neighbors_ann, precursor_tol_mass,
-            precursor_tol_mode, rt_tol, distances, indices, indptr, indptr_i)
-        index.reset()
-        indptr_i += n_split
-    if len(identifiers) == 0:
-        return pd.DataFrame(columns=['filename', 'spectrum_id', 'precursor_mz',
-                                     'retention_time'])
+
+    # Figure out a decent value for the n_list hyperparameter based on
+    # the number of vectors.
+    # Rules of thumb from the Faiss wiki:
+    # https://github.com/facebookresearch/faiss/wiki/Guidelines-to-choose-an-index#how-big-is-the-dataset
+    if n_spectra < 10**2:
+        # Use a brute-force index instead of an ANN index when there
+        # are only a few items.
+        n_list = -1
+    elif n_spectra < 10**6:
+        n_list = 2 ** math.floor(math.log2(n_spectra / 39))
+    elif n_spectra < 10**7:
+        n_list = 2**16
+    elif n_spectra < 10**8:
+        n_list = 2**18
+    else:
+        n_list = 2**20
+        if n_spectra > 10**9:
+            logger.warning(
+                "More than 1B vectors to be indexed, consider "
+                "decreasing the ANN size"
+            )
+    if n_list > 0:
+        # Train on random subset of data
+        # https://github.com/facebookresearch/faiss/issues/2047
+        sample_size = min(50 * n_list, n_spectra)
+    else:
+        sample_size = n_spectra
+    train_spectra = dataset.sample(sample_size).to_pandas()
+    train_vectors = np.stack(train_spectra["vector"].values)
+    dim = train_vectors.shape[1]
+    # Create an ANN index using the inner product (proxy for cosine
+    # distance) for fast NN queries.
+    if n_list <= 0:
+        index = faiss.IndexIDMap(faiss.IndexFlatIP(dim))
     else:
-        return pd.DataFrame({'filename': filenames,
-                             'spectrum_id': identifiers,
-                             'precursor_mz': np.hstack(precursor_mzs),
-                             'retention_time': np.hstack(rts)})
+        index = faiss.IndexIVFFlat(
+            faiss.IndexFlatIP(dim), dim, n_list, faiss.METRIC_INNER_PRODUCT
+        )
+        index.nprobe = min(math.ceil(index.nlist / 8), n_probe)
+    # Compute cluster centroids.
+    # noinspection PyArgumentList
+    index.train(train_vectors)
+
+    # Add the vectors to the index in batches.
+    batch_start = 0
+    with tqdm(desc="Batches added to index", unit="batch") as pbar:
```

Comment:
ah yes, will change it back.
btw, I tried many things to get an actual progress bar (now it just shows the iteration counter), but it doesn't really work for an iterator unless you know the total number of batches. Because lance does weird stuff and ignores the batch size I can't really set the total number of batches.

Reply:
A bit unfortunate, but just having an iteration counter is better than nothing.